trace_decoder/src/observer.rs

```diff
@@ -0,0 +1,110 @@
+use std::collections::BTreeMap;
+use std::marker::PhantomData;
+
+use ethereum_types::H256;
+
+use crate::core::IntraBlockTries;
+use crate::typed_mpt::{ReceiptTrie, StorageTrie, TransactionTrie};
+
+/// Observer API for the trace decoder.
+/// Observer is used to collect various debugging and metadata info
+/// from the trace decoder run.
+pub trait Observer<StateTrieT: Clone> {
+    /// Collect tries after the transaction/batch execution.
+    ///
+    /// Passing the arguments one by one through reference, because
+    /// we don't want to clone argument tries in case they are not used in
+    /// observer.
+    fn collect_tries(
+        &mut self,
+        block: u64,
+        batch: u64,
+        state_trie: &StateTrieT,
+        storage: &BTreeMap<H256, StorageTrie>,
+        transaction_trie: &TransactionTrie,
+        receipt_trie: &ReceiptTrie,
+    );
+}
+
+#[derive(Debug)]
+/// Tries observer collected data element.
+pub struct TriesObserverElement<StateTrieT> {
+    /// Block where the tries are collected.
+    pub block: u64,
+    /// Tries were collected after every batch execution.
+    pub batch: u64,
+    /// Collected tries.
+    pub tries: IntraBlockTries<StateTrieT>,
+}
+
+/// Observer for collection of post-execution tries from the
+/// trace decoder run.
+#[derive(Debug)]
+pub struct TriesObserver<StateTrieT> {
+    /// Collected data in the observer pass
+    pub data: Vec<TriesObserverElement<StateTrieT>>,
+}
+
+impl<StateTrieT> TriesObserver<StateTrieT> {
+    /// Create new tries collecting observer.
+    pub fn new() -> Self {
+        TriesObserver::<StateTrieT> { data: Vec::new() }
+    }
+}
+
+impl<StateTrieT: Clone> Observer<StateTrieT> for TriesObserver<StateTrieT> {
+    fn collect_tries(
+        &mut self,
+        block: u64,
+        batch: u64,
+        state_trie: &StateTrieT,
+        storage: &BTreeMap<H256, StorageTrie>,
+        transaction_trie: &TransactionTrie,
+        receipt_trie: &ReceiptTrie,
+    ) {
+        self.data.push(TriesObserverElement {
+            block,
+            batch,
+            tries: IntraBlockTries {
+                state: state_trie.clone(),
+                storage: storage.clone(),
+                transaction: transaction_trie.clone(),
+                receipt: receipt_trie.clone(),
+            },
+        });
+    }
+}
+
+impl<StateTrieT> Default for TriesObserver<StateTrieT> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Dummy observer which does not collect any data.
+#[derive(Default, Debug)]
+pub struct DefaultObserver<StateTrieT> {
+    phantom: PhantomData<StateTrieT>,
+}
```

Comment:
```suggestion
#[derive(Default, Debug)]
pub struct DummyObserver<StateTrieT> {
    phantom: PhantomData<StateTrieT>,
}
```

Reply:
Clippy would yell that you do not have comment for public structure.